pkg/observable/channel/replay_test.go

```diff
@@ -42,46 +56,135 @@ func TestReplayObservable(t *testing.T) {
-			if v != expected {
-				t.Errorf("Expected value %d, but got %d", expected, v)
-			}
+			require.Equal(t, expected, v)
 		case <-time.After(1 * time.Second):
-			t.Errorf("Did not receive expected value %d in time", expected)
+			t.Fatalf("Did not receive expected value %d in time", expected)
 		}
 	}
 }
 
-func TestReplayObservable_Next(t *testing.T) {
-	var n, ctx = 3, context.Background()
+func TestReplayObservable_Last_Full_ReplayBuffer(t *testing.T) {
+	values := []int{1, 2, 3, 4, 5}
+	tests := []struct {
+		name             string
+		replayBufferSize int
+		// lastN is the number of values to return from the replay buffer
+		lastN          int
+		expectedValues []int
+	}{
+		{
+			name:             "n < replayBufferSize",
+			replayBufferSize: 5,
+			lastN:            3,
+			// the replay buffer is not full so Last should return values
+			// starting from the first published value.
+			expectedValues: values[:3], // []int{1, 2, 3},
+		},
+		{
+			name:             "n = replayBufferSize",
+			replayBufferSize: 5,
+			lastN:            5,
+			expectedValues:   values,
+		},
+		{
+			name:             "n > replayBufferSize",
+			replayBufferSize: 3,
+			lastN:            5,
+			// the replay buffer is full so Last should return values starting
+			// from lastN - replayBufferSize.
+			expectedValues: values[2:], // []int{3, 4, 5},
+		},
+	}
 
-	replayObsvbl, publishCh := channel.NewReplayObservable[int](ctx, n)
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			var ctx = context.Background()
 
-	values := []int{1, 2, 3, 4, 5}
-	for _, value := range values {
+			replayObsvbl, publishCh :=
+				channel.NewReplayObservable[int](ctx, tt.replayBufferSize)
+
+			for _, value := range values {
+				publishCh <- value
+				time.Sleep(time.Millisecond)
+			}
+
+			actualValues := replayObsvbl.Last(ctx, tt.lastN)
+			require.ElementsMatch(t, tt.expectedValues, actualValues)
+		})
+	}
+}
+
+func TestReplayObservable_Last_Blocks_Goroutine(t *testing.T) {
+	var (
+		lastN    = 5
+		splitIdx = 3
+		values   = []int{1, 2, 3, 4, 5}
+		ctx      = context.Background()
+	)
+
+	replayObsvbl, publishCh := channel.NewReplayObservable[int](ctx, lastN)
+
+	// Publish values up to splitIdx.
+	for _, value := range values[:splitIdx] {
 		publishCh <- value
 		time.Sleep(time.Millisecond)
 	}
 
-	require.ElementsMatch(t, []int{3}, replayObsvbl.Last(ctx, 1))
-	require.Equal(t, []int{3, 4}, replayObsvbl.Last(ctx, 2))
-	require.Equal(t, []int{3, 4, 5}, replayObsvbl.Last(ctx, 3))
+	require.ElementsMatch(t, []int{1}, replayObsvbl.Last(ctx, 1))
+	require.ElementsMatch(t, []int{1, 2}, replayObsvbl.Last(ctx, 2))
+	require.ElementsMatch(t, []int{1, 2, 3}, replayObsvbl.Last(ctx, 3))
+
+	// Concurrently call Last with a value greater than the replay buffer size.
+	lastValues := make(chan []int, 1)
+	go func() {
+		// Last should block until lastN values have been published.
+		lastValues <- replayObsvbl.Last(ctx, lastN)
+	}()
+
+	select {
```

Comment:
// Since `lastN` is greater than than number of values we have published to `publishCh`, the
// call to `Last(ctx, LastN)` above in the goroutine is expected to block, and therefore this select
// statement should time out.

Reply:
I rearranged some things but didn't end up using this verbatim. LMK if you would like me to take another pass at it.

Reply:
Resolving this thread and will re-review.